photon-api/src/main/scala/com/linkedin/photon/ml/optimization/RandomEffectOptimizationTracker.scala

```diff
@@ -89,57 +97,56 @@ protected[ml] class RandomEffectOptimizationTracker(val optimizationStatesTracke
    *
    * @return A map of convergence reason to its count of occurrence from the random effect optimization problems
    */
-  def countConvergenceReasons: Map[String, Int] = {
-    optimizationStatesTrackers.map { optimizationStatesTracker =>
-      if (optimizationStatesTracker.convergenceReason.isDefined) {
-        (optimizationStatesTracker.convergenceReason.get.reason, 1)
-      } else {
-        (RandomEffectOptimizationTracker.NOT_CONVERGED, 1)
-      }
-    }.reduceByKey(_ + _).collectAsMap()
-  }
+  def countConvergenceReasons: Map[String, Int] = optimizationStatesTrackers.map { optimizationStatesTracker =>
+      (optimizationStatesTracker
+        .convergenceReason
+        .map(_.reason)
+        .getOrElse(RandomEffectOptimizationTracker.NOT_CONVERGED), 1)
+    }
+    .reduceByKey(_ + _)
+    .collectAsMap()
 
   /**
    * Get stats counter of the number of iterations tracked in all optimization states trackers
    *
    * @return A [[StatCounter]] of the number of iterations tracked in all optimization states trackers
    */
-  def getNumIterationStats: StatCounter = {
-    optimizationStatesTrackers.map(_.getTrackedStates.length).stats()
-  }
+  def getNumIterationStats: StatCounter = optimizationStatesTrackers.map(_.getTrackedStates.length).stats()
 
   /**
    * Get stats counter on the elapsed time tracked in all optimization states trackers
    *
    * @return A [[StatCounter]] of the elapsed time tracked in all optimization states trackers
    */
-  def getElapsedTimeStats: StatCounter = {
-    optimizationStatesTrackers.map { optimizationStatesTracker =>
+  def getElapsedTimeStats: StatCounter = optimizationStatesTrackers.map { optimizationStatesTracker =>
       val timeHistory = optimizationStatesTracker.getTrackedTimeHistory
+
       if (timeHistory.length > 0) {
         timeHistory.last * 1e-6
       } else {
         0.0
       }
-    }.stats()
-  }
+    }
+    .stats()
 
   /**
    * Build a summary string for the tracker
    *
    * @return string representation
    */
   override def toSummaryString: String = {
+
     val convergenceReasons = countConvergenceReasons
     val numIterationsStats = getNumIterationStats
     val timeElapsedStats = getElapsedTimeStats
+
     RandomEffectOptimizationTracker.SUMMARY_FORMAT.format(convergenceReasons, numIterationsStats, timeElapsedStats)
   }
 }
 
 object RandomEffectOptimizationTracker{
-  protected[optimization] val NOT_CONVERGED = "Not converged"
 
+  protected[optimization] val NOT_CONVERGED = "Not converged"
```

Comment:
We need to analyze the possible reasons of non-convergence to tackle them at the root as much as possible (just a reminder, let's keep a todo item about that somewhere).